Implement partial auroc metric

prampey opened the pull request:
## What does this PR do?
Working implementation for #3675 
Some notes:
- Wrote very basic tests. Need some advice for more extensive testing.
- Commented the ```@auc_decorator``` out because the area had to be corrected after calculation. See comments.
- Perhaps the code can be optimized more to reduce trips between GPU and CPU. Planning to do that once extensive tests are in place.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Ofc ;)

## Resources
- scikit's PR on partial AUC https://github.com/scikit-learn/scikit-learn/issues/3273

pep8speaks commented:
Hello @prampey! Thanks for opening this PR. 

* In the file [`pytorch_lightning/metrics/functional/classification.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py):

> [Line 871:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L871): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 876:22](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L876): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 881:9](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L881): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 882:9](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L882): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 885:37](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L885): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 885:65](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L885): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 886:41](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L886): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 889:9](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L889): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 890:9](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L890): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 898:1](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L898): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 900:5](https://github.com/PyTorchLightning/pytorch-lightning/blob/a4fee98aaab22ad211e0caa4f0499f6db5aa8b06/pytorch_lightning/metrics/functional/classification.py#L900): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)



Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

prampey commented:
I have added tests across the range `(0, 1] `for max_fpr. However, the tests are failing because there is a difference in float precision of `1e-3` popping up between the torch and sklearn results. Here's an example:
```
E       AssertionError: assert False                                                                                                                            
E        +  where False = <built-in method allclose of type object at 0x7fa60c1a3a80>(tensor(0.5120, device='cuda:0'), tensor(0.5122, device='cuda:0'))         
E        +    where <built-in method allclose of type object at 0x7fa60c1a3a80> = torch.allclose                                                                
```
This error originated because I changed the following code from
`pred = torch.randint(n_cls_pred, (300,), device=device)` to `pred = torch.rand((300,), device=device)` . 
The AUC metric is usually applied to prediction probabilities and not prediction labels themselves, hence the need to generate random prediction values between 0 and 1.

I'd appreciate any help on taking this forward :)

ddrevicky commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -867,11 +868,35 @@ def auroc(
+    def _auroc(pred, target, sample_weight, pos_label, max_fpr):
+        if max_fpr is None or max_fpr == 1:
+            fpr, tpr, _ = roc(pred, target, sample_weight, pos_label)
+            return auc(fpr, tpr, reorder=True)
```
```suggestion
            return auc(fpr, tpr, reorder=False)
```

ddrevicky commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -867,11 +868,35 @@ def auroc(
+        fpr = torch.cat([fpr[:stop], max_fpr.view(1)])
+        # Compute partial AUC
+        # Do i need to reorder?
+        partial_auc = auc(fpr, tpr, reorder=True)
```
```suggestion
        partial_auc = auc(fpr, tpr, reorder=False)
```

ddrevicky commented:
Hey @prampey the problem was with the `reorder=True` param to `auc` which uses unstable `torch.argsort` and this can cause the tpr/fpr pairs from which the AUC is calculated to be switched and that leads to computation of the wrong value. I tried setting it to `False` and your tests seem to pass locally :). Good job with the PR.

I think you just need to update the docstring and CHANGELOG now.

justusschock commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -867,11 +868,35 @@ def auroc(
+        # torch.bucketize is slightly more performant than torch.searchsorted
+        # for 1-D arrays
+        # Add a single point at max_fpr and interpolate its tpr value
+        stop = torch.bucketize(max_fpr, fpr, out_int32=True, right=True)
```
This restricts the PyTorch version, right? It was introduced quite recently.

ddrevicky replied:
Both `torch.bucketize` or the alternative`torch.searchsorted` were added in Pytorch 1.6. Would it be better to replace them with manually written code?

justusschock replied:
@ddrevicky I think we can keep it as it currently is and just add a version check in the beginning

prampey replied:
So something like this?
```python
if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
    raise RuntimeError('max_fpr argument requires torch.bucketize which is not available below PyTorch version 1.6')
```

SkafteNicki commented:
Hi @prampey, how is it going here?

prampey commented:
Hi @SkafteNicki , haven't had an opportunity to work on this again. Will hopefully put in some time this week :)

Borda commented:
@prampey can we get this merge by 1.1? 🐰 
cc: @SkafteNicki @justusschock

SkafteNicki commented:
@prampey I have tried updating this PR such that it is ready to be merged. Please confirm that it is still doing what you intended.

prampey commented:
Hey @SkafteNicki , sorry I've been hung up with my job and sending college applications. :see_no_evil: 
Everything seems to be working fine, added a few more values to test out the whole range of (0,1] for `max_fpr  `

SkafteNicki reviewed (dismissed):
LGTM

rohitgr7 commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -553,6 +556,8 @@ def auroc(
         target: ground-truth labels
         sample_weight: sample weights
         pos_label: the label for the positive class
+        max_fpr: If not `None`, calculates standardized partial AUC over the
+            range [0, max_fpr]. Should be a float between 0 and 1 (Default: None)
```
```suggestion
        max_fpr: If not ``None``, calculates standardized partial AUC over the
            range [0, max_fpr]. Should be a float between 0 and 1.
```

rohitgr7 commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -544,6 +546,7 @@ def auroc(
         target: torch.Tensor,
         sample_weight: Optional[Sequence] = None,
         pos_label: int = 1.,
+        max_fpr: float = None
```
```suggestion
        max_fpr: float = None,
```

rohitgr7 commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -569,11 +574,35 @@ def auroc(
+        if max_fpr <= 0 or max_fpr > 1:
+            raise ValueError("Expected max_fpr in range (0, 1], got: %r" % max_fpr)
+        if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+            raise RuntimeError('`max_fpr` argument requires `torch.bucketize` which '
```
```suggestion
            raise RuntimeError('`max_fpr` argument requires `torch.bucketize` which'
```

rohitgr7 commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -569,11 +574,35 @@ def auroc(
+        if max_fpr is None or max_fpr == 1:
+            fpr, tpr, _ = __roc(pred, target, sample_weight, pos_label)
+            return auc(fpr, tpr)
+        if max_fpr <= 0 or max_fpr > 1:
+            raise ValueError("Expected max_fpr in range (0, 1], got: %r" % max_fpr)
```
```suggestion
        if not (isinstance(max_fpr, float) and 0 < max_fpr <= 1):
            raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
```

rohitgr7 reviewed (commented):
just some comments.

Borda commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -569,11 +574,35 @@ def auroc(
-        return _roc(pred, target, sample_weight, pos_label)
-
-    return _auroc(pred=pred, target=target, sample_weight=sample_weight, pos_label=pos_label)
+    def _auroc(pred, target, sample_weight, pos_label, max_fpr):
+        if max_fpr is None or max_fpr == 1:
+            fpr, tpr, _ = __roc(pred, target, sample_weight, pos_label)
+            return auc(fpr, tpr)
+        if not (isinstance(max_fpr, float) and 0 < max_fpr <= 1):
+            raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
+        if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+            raise RuntimeError('`max_fpr` argument requires `torch.bucketize` which'
+                               ' is not available below PyTorch version 1.6')
+
+        fpr, tpr, _ = __roc(pred, target, sample_weight, pos_label)
+        max_fpr = torch.tensor(max_fpr, device=fpr.device)
+        # Add a single point at max_fpr and interpolate its tpr value
+        stop = torch.bucketize(max_fpr, fpr, out_int32=True, right=True)
+        weight = (max_fpr - fpr[stop - 1]) / (fpr[stop] - fpr[stop - 1])
+        interp_tpr = torch.lerp(tpr[stop - 1], tpr[stop], weight)
+        tpr = torch.cat([tpr[:stop], interp_tpr.view(1)])
+        fpr = torch.cat([fpr[:stop], max_fpr.view(1)])
+
+        # Compute partial AUC
+        partial_auc = auc(fpr, tpr)
+
+        # McClish correction: standardize result to be 0.5 if non-discriminant
+        # and 1 if maximal
+        min_area = 0.5 * max_fpr**2
+        max_area = max_fpr
+        return 0.5 * (1 + (partial_auc - min_area) / (max_area - min_area))
```
why is this added to deprecated func?

SkafteNicki replied:
it is not a deprecated func, it is a function (`_auroc`) inside a function (`auroc`). Originally there was a decorator in front of inner function, but that has been removed. Therefore, we unwrap the content of the inner function.

SkafteNicki replied:
@Borda I have fixed it now so it should be more clear that it is not a deprecated func

Borda commented on pytorch_lightning/metrics/functional/classification.py:
```diff
@@ -569,11 +573,32 @@ def auroc(
+
+    # McClish correction: standardize result to be 0.5 if non-discriminant
+    # and 1 if maximal
+    min_area = 0.5 * max_fpr**2
```
```suggestion
    min_area = 0.5 * max_fpr ** 2
```

Borda commented on tests/metrics/functional/test_classification.py:
```diff
@@ -197,18 +198,50 @@ def test_binary_clf_curve(sample_weight, pos_label, exp_shape):
+@pytest.mark.skipif(LooseVersion(torch.__version__) < LooseVersion('1.6.0'),
+                    reason='requires torch v1.6 or higher to test max_fpr argument')
+@pytest.mark.parametrize(['max_fpr'], [
+    pytest.param(None),
+    pytest.param(1),
+    pytest.param(0.99),
+    pytest.param(0.9),
+    pytest.param(0.75),
+    pytest.param(0.5),
+    pytest.param(0.25),
+    pytest.param(0.1),
+    pytest.param(0.01),
+    pytest.param(0.001),
```
```suggestion
    None, 1, 0.99, 0.9, 0.75, 0.5, 0.25, 0.1, 0.01, 0.001,
```
